src/main/java/htsjdk/samtools/BAMFileReader.java

```diff
@@ -772,9 +772,11 @@ public SAMRecord next() {
      * Starting point of iteration is wherever current file position is when the iterator is constructed.
      */
     private class BAMFileIterator extends AbstractBamIterator {
-        private SAMRecord mNextRecord = null;
+        private BamRecordDecodingInfo mNextRecord = null;
         private final BAMRecordCodec bamRecordCodec;
-        private long samRecordIndex = 0; // Records at what position (counted in records) we are at in the file
+        private final AsyncBamDecoder mAsync;
+        private final ConcurrentLinkedQueue<BAMRecordCodec> mAvailableIdleCodec;
```

Comment:
I poked at this a bit because I don't like having separate `BAMRecordCodec` and `mAvailableIdleCodec` fields.  I think it can be cleaned up by using an ObjectPool<BAMRecordCodec>.
Here's a simple implementation that I believe works (although I haven't tested it yet).

```
public interface ObjectPool<T> {
    <U> U borrow(Function<T, U> computeWithPoolObject);
}

public class QueuePool<T> implements ObjectPool<T> {
    private final Queue<T> queue = new ConcurrentLinkedQueue<>();
    private final Supplier<T> objectSupplier;

    public QueuePool(Supplier<T> objectSupplier) {
        this.objectSupplier = objectSupplier;
    }

    @Override
    public <U> U borrow(Function<T,U> usePoolObject) {
        T object = queue.poll();
        try {
            if(object == null){
                object = objectSupplier.get();
            }
            return usePoolObject.apply(object);
        } finally {
            queue.offer(object);
        }
    }
}

public class SingletonPool<T> implements ObjectPool<T> {
    private final T object;

    SingletonPool(T object){
        this.object = object;
    }

    @Override
    public synchronized <U> U borrow(final Function<T, U> usePoolObject) {
        return usePoolObject.apply(object);
    }
}
```

The two fields can be merged into a single ObjectPool that's instantiated either as a singleton or a queue. 

I think there are a bunch of places in BlockCompressedInputStream and AsyncBlockCompressedInputStream that could be significantly simplified using something like this as well.  It may need more complicated new object generation though, I haven't worked through it yet.

Reply:
> I don't like having separate BAMRecordCodec and mAvailableIdleCodec fields.

bamRecordCodec is the codec attached to the stream and is the only codec used in sync calls. In async mode, the stream codec is used to read the record length from the stream, and the available codecs are recycled for parallel record parsing. They're both needed because they do different things.

One could add the stream codec to the pool and reattached to the stream every time you needed to read the record length but that's even less intuitive than what I have now. I've renamed the variables so it's clear what they do.

Reply:
>I think there are a bunch of places in BlockCompressedInputStream and AsyncBlockCompressedInputStream that could be significantly simplified using something like this as well. It may need more complicated new object generation though, I haven't worked through it yet.

How do you like my RecyclingSupplier implementation? AsyncBlockCompressedInputStream.mFreeDecompressedBlockedBuffers doesn't fit the model due to the potential for different buffer sizes so that's been kept as a ConcurrentLinkedQueue but the others have been simplified.

Reply:
Decided to just remove the codec cache entirely. BamRecordCodec is pretty cheap to clone and it removes a per record (not per batch) source of synchronisation